src/dune_rpc_impl/server.ml

```diff
@@ -204,13 +205,28 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
     Handler.implement_notification rpc Procedures.Public.shutdown shutdown
   in
   let () =
+    let svar = Build_system.progress () in
+    let module Poll_comparable = Comparable.Make (Poller) in
+    let module Map = Poll_comparable.Map in
+    let pollers = ref Map.empty in
     Handler.implement_poll rpc Procedures.Poll.progress
       ~on_cancel:(fun _session poller ->
         let p = Long_poll.progress (Fdecl.get t).long_poll in
         Long_poll.Instance.client_cancel p poller)
       ~on_poll:(fun _session poller ->
-        let p = Long_poll.progress (Fdecl.get t).long_poll in
-        Long_poll.Instance.poll p poller)
+        let last = Map.find !pollers poller in
+        let until =
+          match last with
+          | None -> fun _ -> true
+          | Some prev -> Build_system.Progress.equal prev
+        in
+        let+ (next : Build_system.Progress.t) = Fiber.Svar.wait svar ~until in
```

Comment:
I guess this also shows another issue with `~until`. There's no good way express "wake up me on the next write" using a predicate. You need to do something funky like:

```ocaml
let until =
  let skip = ref true in
  fun _ ->
    skip := not !skip;
    !skip
```

So instead, I'm now saving the last progress I've reported to and making sure I don't report duplicate progress consecutively. That seems a little more complicated than necessary.

Reply:
Why do you want to be awaken on the next write?

Reply:
It will simplify the implementation and there will be downsides to the user.

Reply:
I see what you mean. However, if we could say "wake me up on the next write", we would also need to answer the question: "was there a write since the last time this particular poller made a request?". So we would still need to remember a piece of information for each poller. i.e. we would still need this `pollers` map. It would just hold a value of a different type and the code would be pretty much the same.

The current code seems clear to me: for each poller, we remember the last value they observed.

Reply:
BTW, it seems to me that this logic shouldn't have to be replicated for each king of long polling request. Instead, there should be something that takes an `'a Svar.t`, an `'a` equality function and constructs a long polling request.

Reply:
Yes, I agree. I'm just making sure it works before rewriting everything in a generic way.